### ISSUE TITLE
Obsidian freezes for 10-15 seconds when pressing map view 

### ISSUE THREAD

**Issue by zeltak** [esm7/obsidian-map-view#19]
thx for the fix!

I updated to the latest version

unfortunately it seems now on my system pressing the map view freezes obsidian for a good 10-15 seconds :)

plus it seems that didn't fix the world view 

as always happy to help debug!

best

Z

_Originally posted by @zeltak in https://github.com/esm7/obsidian-map-view/issues/14#issuecomment-890802338_

**Comment by zeltak** [comment]
see above comments

**Comment by esm7** [comment]
1. Are you sure you upgraded the plugin version to 0.0.9?
2. Can you post your configuration again?
3. Is there something showing in the development console (Ctrl+Shift+I) when you open the plugin or work with it?

**Comment by zeltak** [comment]
Hi again

> Are you sure you upgraded the plugin version to 0.0.9?
![image](https://user-images.githubusercontent.com/498410/128344890-6cda1b89-ba9e-4980-9508-5fa2a7317820.png)

> Can you post your configuration again?
sure attached
[data.json.zip](https://github.com/esm7/obsidian-map-view/files/6938353/data.json.zip)
 

> Is there something showing in the development console (Ctrl+Shift+I) when you open the plugin or work with it?

this time obsidian frooze completely :) here is the dev view

![image](https://user-images.githubusercontent.com/498410/128345734-9e5ee3d5-e777-41e8-a851-fd499a01c36b.png)


thx again

itai

**Comment by esm7** [comment]
Sorry I should have been more detailed :) Under the dev view there's a "console" tab (it's at the bottom of your screenshot). Can you check if it shows anything?

**Comment by zeltak** [comment]
No worries , happy to debug:)

here it is (this time obsidian froze for around 15 seconds):

![image](https://user-images.githubusercontent.com/498410/128352544-eba2d7d3-8224-4373-859d-8e630baece58.png)

**Comment by esm7** [comment]
We're not as lucky this time, I'm unable to recreate it.
A few more things if you don't mind...
1. Can you try to disable all community plugins except Map View, restart Obsidian and see if this makes a difference? If so, it would be great if you can locate which plugin is interfering.
2. After that initial freeze, if you click on a note then go back to the map, does it freeze again?
3. You mentioned you still have the "world view" issue. Do you mean that when you click a note, then click "show on map", you get the world view rather than a zoom on the location? Does it happen with all locations?

**Comment by zeltak** [comment]
> 1. Can you try to disable all community plugins except Map View, restart Obsidian and see if this makes a difference? If so, it would be great if you can locate which plugin is interfering.

did that and still get eh freeze :D 

> 2. After that initial freeze, if you click on a note then go back to the map, does it freeze again?

i cant even click on a note obsidian is completely frozen

> 3. You mentioned you still have the "world view" issue. Do you mean that when you click a note, then click "show on map", you get the world view rather than a zoom on the location? Does it happen with all locations?

yes exactly and yes it happens in all locations (tried 7-8 different notes and locations )

thx so much !

Z

**Comment by esm7** [comment]
I just released a 0.1.0 version and added a Debug Logs toggle in the settings. Can you update, turn that on, reload the plugin, let it sit until it's done whatever it is doing, and then send me the same console output like above?

**Comment by HEmile** [comment]
As another point of reference, I have the same problem, it freezes for about 10 seconds **after** it prints out the updateMapToState:

![image](https://user-images.githubusercontent.com/1289165/128410950-ea15cf92-5033-4e80-b523-11d16e93f208.png)

**Comment by zeltak** [comment]
> I just released a 0.1.0 version and added a Debug Logs toggle in the settings. Can you update, turn that on, reload the plugin, let it sit until it's done whatever it is doing, and then send me the same console output like above?

sure :) updated , turned in the advanced debug and"

-  first time obsidian the map didn't show up and i got this error

![image](https://user-images.githubusercontent.com/498410/128474479-381212af-2ad3-4c89-afaa-42746d5945f8.png)

- second time obsidian just completely froze up:

![image](https://user-images.githubusercontent.com/498410/128475139-2e03ffb1-d551-408d-a90f-4f508a44fd3b.png)


am i doing something wrong :D

best 

Z

**Comment by esm7** [comment]
Thanks for the help so far! I have a suspicion of what might be going on and added more debug logs to catch it.
@HEmile, @zeltak, do you mind -
1. Unzipping the attached debug build of the plugin into VAULT/.obsidian/plugins/obsidian-map-view (make sure there's already a main.js file there before unzipping so you'll be sure you're in the right place),
2. Doing the same thing, again with debug logs turned on?
[main.zip](https://github.com/esm7/obsidian-map-view/files/6945268/main.zip)

**Comment by HEmile** [comment]
Looks like the setState phase is very expensive:

![image](https://user-images.githubusercontent.com/1289165/128508572-e3b73c27-bb19-406a-abb7-3bcc730e71fe.png)

**Comment by esm7** [comment]
Ok I think I got it! I'm guessing from the log that Obsidian on your system doesn't like plugins to `await` while setting a state, regardless of the actual action (`updateMapToState`) not being very long.
Can you try the attached fix and let me know if it solves the issue?
[main.zip](https://github.com/esm7/obsidian-map-view/files/6945460/main.zip)

**Comment by HEmile** [comment]
It's now giving this, but it's still freezing actually...

![image](https://user-images.githubusercontent.com/1289165/128512783-ae7cd140-4886-46f7-b3d0-a245290a0ff6.png)

**Comment by zeltak** [comment]
Hi again

i also tried it and its still freezing for me :) 

![image](https://user-images.githubusercontent.com/498410/128633589-4d4b3539-c8a6-4f4c-9770-035c955d3e9a.png)



happy to further debug

Z

**Comment by esm7** [comment]
For both of you guys the performance counters are fundamentally different than mine. I made an optimization that might help and added some more specific printouts.
Would appreciate another run.
[main.zip](https://github.com/esm7/obsidian-map-view/files/6950959/main.zip)

**Comment by zeltak** [comment]
sure

here is the output with the latest version you sent 

Still took about 2 minutes to load and again the map is zoomed out

![image](https://user-images.githubusercontent.com/498410/128699560-95069611-772f-4402-8fd4-dbb6912545ee.png)

thx for looking into this!

Z

**Comment by esm7** [comment]
I'm sorry for disappearing, I didn't have any time to work on this project for more than a week. This bug is my top priority, and I'll devote the little time I have in the next couple of days to getting it sorted.
I suspect the freeze started in version 0.0.9, so I inspected the diff between the versions and came up with another attempt. Can you try the attached?
[main.zip](https://github.com/esm7/obsidian-map-view/files/7006586/main.zip)

**Comment by zeltak** [comment]
thx!

no worries @esm7 :)

i tried the latest it dosent see to give any errors but still the same super slow response :)

![image](https://user-images.githubusercontent.com/498410/129892087-4b0744a6-005b-42aa-91cc-1ecf682d0b49.png)


best

Z

**Comment by esm7** [comment]
You seem to have over 1400 location markers, is this correct or the result of some parsing error?
That's very possibly the key for why it doesn't recreate for me. How do these markers look like, are they mostly in separate notes or is it a few notes with many inline markers?
Also, is my understanding correct that the problem started in version 0.0.9 of the plugin?
If your inline locations are still backwards-compatible (you didn't convert them to the `geo:...` syntax), can you check if in version 0.0.8 (attached) the freeze is not there?
[main.zip](https://github.com/esm7/obsidian-map-view/files/7006963/main.zip)

**Comment by zeltak** [comment]
Hi again

1400 seems to much but i may have a few hundreds across my notes

its always in the yaml part such as

```
---
tags: [ nyc]
status: 5
rating: 3
location: [40.784240722656,-73.948684692383]
counter: 
---
# Pata Market
```


each note has a location entry (i have a few hundreds)

i don't think i ever had 0.0.8, im a neweb in obsidian :)

thx again

Z

**Comment by esm7** [comment]
@HEmile was there a version upgrade that seems to have worsened the situation or was it always like that with your amount of markers?
It could be "just" the inherent issue that it takes a long time to parse thousands of files and build markers for them, and caching is needed, as I'm planning anyway to do for [this](https://github.com/esm7/obsidian-map-view/issues/17) and [this](https://github.com/esm7/obsidian-map-view/issues/13).
Either way, I wrote a script that generates 1000 notes with location data and the problem now recreates, so I can investigate is there is a bottleneck that can be tackled before I'm done with caching and clusters.

**Comment by esm7** [comment]
After further investigation seems like parsing and creating markers for thousands of notes is actually very fast, and all the time goes to add these markers to Leaflet.js.
I'll investigate if there is a faster way to batch-add markers, but it's possible that only [clusters](https://github.com/esm7/obsidian-map-view/issues/13) can handle these quantities.

**Comment by HEmile** [comment]
> @HEmile was there a version upgrade that seems to have worsened the situation or was it always like that with your amount of markers?
I don't think there was an earlier version that was quicker, no. Like you mention, it's more likely related to needing to either cache or with creating the markers. Like mentioned, clusters would be ideal for this! (In fact, that's a feature I'm quite missing in terms of usability too. It can look quite cluttered! I hope it's not too hard to implement...)

**Comment by esm7** [comment]
It doesn't look hard at all, after fixing a few small bugs this is definitely the next thing to do. It will also greatly help make the plugin more usable on mobile.

**Comment by esm7** [comment]
It's finally here, marker clusters are now available on version 0.2.1, and I'm curious to know how this works for you guys with many markers!

**Comment by HEmile** [comment]
Oh wow this is excellent! So smooth and quick :) Very amazing work!! 
![image](https://user-images.githubusercontent.com/1289165/138239145-22b617ff-78fe-45af-850f-cd422c4302a8.png)
